src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,755 @@
+	struct dirent     *dentries = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "LsIOOI", &hdl, &path, &anchor_index, &files, &dirs,
+				       &readdir_chunk);
+
+	assert(hdl->dfs != NULL);
+
+	if (readdir_chunk == 0) {
+		return PyLong_FromLong(EINVAL);
+	}
+
+	D_ALLOC_ARRAY(dentries, readdir_chunk);
+	if (dentries == NULL) {
+		rc = ENOMEM;
+		goto out;
+	}
+
+	D_ALLOC_ARRAY(stats, readdir_chunk);
+	if (stats == NULL) {
+		rc = ENOMEM;
+		goto out;
+	}
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup object at '%s': %s (rc=%d)", path, strerror(rc), rc);
+		goto out;
+	}
+
+	daos_anchor_init(&anchor, 0);
+	rc = dfs_obj_anchor_set(obj, anchor_index, &anchor);
+	if (rc) {
+		D_ERROR("Could not set anchor '%u' for object at '%s': %s (rc=%d)", anchor_index,
+			path, strerror(rc), rc);
+		goto out;
+	}
+
+	do {
+		nr = readdir_chunk;
+		rc = dfs_readdirplus(hdl->dfs, obj, &anchor, &nr, dentries, stats);
+		if (rc) {
+			D_ERROR("Readdirplus of '%s' failed: %s (rc=%d)", path, strerror(rc), rc);
+			goto out;
+		}
+
+		for (uint32_t i = 0; i < nr; ++i) {
+			const struct stat *st    = &stats[i];
+			PyObject          *dname = PyUnicode_FromString(dentries[i].d_name);
+			if (dname == NULL) {
+				rc = ENOMEM;
+				goto out;
+			}
+
+			if (S_ISDIR(st->st_mode)) {
+				rc = PyList_Append(dirs, dname);
+				Py_DECREF(dname); /* PyList_Append does not steal the reference */
+				if (rc) {
+					goto out;
+				}
+				continue;
+			}
+
+			PyObject *item = PyTuple_New(2);
+			if (item == NULL) {
+				rc = ENOMEM;
+				goto out;
+			}
+
+			PyTuple_SetItem(item, 0, dname); /* steals the reference */
+			PyTuple_SetItem(item, 1, PyLong_FromLong(st->st_size));
+
+			rc = PyList_Append(files, item);
+			Py_DECREF(item); /* PyList_Append does not steal the reference */
+			if (rc < 0) {
+				rc = EIO;
+				goto out;
+			}
+		}
+	} while (nr > 0);
+
+out:
+	D_FREE(dentries);
+	D_FREE(stats);
+
+	if (obj) {
+		rc = dfs_release(obj);
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+static PyObject *
+__shim_handle__torch_read(PyObject *self, PyObject *args)
+{
+	ssize_t            rc     = 0;
+	struct dfs_handle *hdl    = NULL;
+	char              *path   = NULL;
+	dfs_obj_t         *obj    = NULL;
+	PyObject          *buffer = NULL;
+	Py_buffer          bview;
+	d_iov_t            iov;
+	daos_size_t        read = 0;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "LsO", &hdl, &path, &buffer);
+	assert(hdl->dfs != NULL);
+
+	if (!PyObject_CheckBuffer(buffer)) {
+		PyErr_SetString(PyExc_TypeError,
+				"Expected an object that supports the buffer protocol");
+		return NULL;
+	}
+
+	if (PyObject_GetBuffer(buffer, &bview, PyBUF_WRITE) == -1) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not writable");
+		return NULL;
+	}
+
+	/*
+	  Since python can use buffer like objects that might not have contiguous memory layout,
+	  let's put a guardrail accepting only buffers with contiguous memory region
```

Comment:
what is the challenge of supporting non-contiguous mem layout? DAOS supports that with the sgl

Reply:
There's no challange, the python wrapper allocates buffer before the call to shim layer: https://github.com/daos-stack/daos/pull/15475/files/92e30a337949e2eb8116cf91476e18c1bd67cec3#diff-2ea6a1bbd40194ac265357a571f9323b1097ff49ef7d887ae663170a9396379dR359 so this is more a guardrail to keep things simple.